src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,384 @@
+            try
+            {
+                int waitResult;
+
+                SynchronizationContext context = SynchronizationContext.Current;
+                if (context != null && context.IsWaitNotificationRequired())
+                {
+                    var handles = new[] { waitHandle.DangerousGetHandle() };
+                    waitResult = context.Wait(handles, false, millisecondsTimeout);
+                    GC.KeepAlive(handles);
+                }
+                else
+                {
+                    waitResult = WaitOneCore(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+                }
+
+                if (waitResult == WaitAbandoned)
+                {
+                    throw new AbandonedMutexException();
+                }
+
+                return waitResult != WaitTimeout;
+            }
+            finally
+            {
+                if (success)
+                    waitHandle.DangerousRelease();
+            }
+        }
+
+        /// <summary>
+        /// Obtains all of the corresponding safe wait handles and adds a ref to each. Since the <see cref="SafeWaitHandle"/>
+        /// property is publically modifiable, this makes sure that we add and release refs one the same set of safe wait
+        /// handles to keep them alive during a multi-wait operation.
+        /// </summary>
+        private static void ObtainSafeWaitHandles(
+            ReadOnlySpan<WaitHandle> waitHandles,
+            out SafeWaitHandle[] safeWaitHandles,
+            out IntPtr[] unsafeWaitHandles)
+        {
+            Debug.Assert(waitHandles != null);
+            Debug.Assert(waitHandles.Length > 0);
+            Debug.Assert(waitHandles.Length <= MaxWaitHandles);
+
+            safeWaitHandles = new SafeWaitHandle[waitHandles.Length];
+            unsafeWaitHandles = new IntPtr[waitHandles.Length];
+            bool success = false;
+            try
+            {
+                for (int i = 0; i < waitHandles.Length; ++i)
+                {
+                    WaitHandle waitHandle = waitHandles[i];
+                    if (waitHandle == null)
+                    {
+                        throw new ArgumentNullException("waitHandles[" + i + ']', SR.ArgumentNull_ArrayElement);
+                    }
+
+                    SafeWaitHandle safeWaitHandle = waitHandle._waitHandle;
+                    if (safeWaitHandle == null)
+                    {
+                        // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                        throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+                    }
+
+                    safeWaitHandle.DangerousAddRef(ref success);
+                    safeWaitHandles[i] = safeWaitHandle;
+                    unsafeWaitHandles[i] = safeWaitHandle.DangerousGetHandle();
+                }
+                success = true;
```

Comment:
Wouldn't usual try/catch instead of success flag work better (the one in DangerousAddRef seems oddly placed anyway) ?

Reply:
I still think try/catch would be better than try/finally with a failed flag

Reply:
Sorry, I misunderstood your comment. I didn't realise this is the one place where the `finally` is really just an error handling. I'll fix it.